lib/api.js

```diff
@@ -0,0 +1,124 @@
+var base      = require('taskcluster-base');
+var github    = require('./github');
+var _         = require('lodash');
+
+// Common schema prefix
+var SCHEMA_PREFIX_CONST = 'http://schemas.taskcluster.net/github/v1/';
+
+// Convert GitHub event types to their expected publisher name
+function eventTypeToPublisherName(eventType) {
+  let firstUnderscore = eventType.indexOf("_")
+  if (firstUnderscore < 0) {
+    return eventType
+  }
+  let publisherName = eventType.substring(0, firstUnderscore)
+  publisherName += eventType.substring(
+    firstUnderscore + 1, firstUnderscore + 2).toUpperCase()
+  publisherName += eventType.substring(firstUnderscore + 2, eventType.length)
+  return eventTypeToPublisherName(publisherName)
+}
+
+/** API end-point for version v1/
+ *
+ * In this API implementation we shall assume the following context:
+ * {
+ *   publisher:      // publisher from base.Exchanges
+ * }
+ */
+var api = new base.API({
+  title:        "TaskCluster GitHub API Documentation",
+  description: [
+    "The github service, typically available at",
+    "`github.taskcluster.net`, is responsible for publishing pulse",
+    "messages in response to GitHub events.",
+    "",
+    "This document describes the API end-point for consuming GitHub",
+    "web hooks"
+  ].join('\n')
+});
+
+// Export API
+module.exports = api;
+
+/** Define tasks */
+api.declare({
+  method:     'post',
+  route:      '/github',
+  name:       'githubWebHookConsumer',
+  scopes:     undefined,
+  title:      "Consume GitHub WebHook",
+  description: [
+    "Capture a GitHub event and publish it via pulse, if it's a push",
+    "or pull request."
+  ].join('\n')
+}, async function(req, res) {
+  let eventType = req.headers['x-github-event']
+  if (!eventType) {
+    res.status(400).send("Missing X-GitHub-Event")
+  }
+
+  let body = req.body
+  if (!body) {
+    req.status(400).send("Request missing a body")
+  }
+
+  // When pulse is activated, locate valid publishers by naming
+  // convention and fail if we don't find one which matches the
+  // event type
+  let publisher = this.publisher[eventTypeToPublisherName(eventType)]
```

Comment:
This pattern generally won't work... Because you change the context of the method.
It works here because of how methods on `this.publisher` is constructor..

Also calling the right hand side `publisher` is a bad idea. Why not do:
```js
var exchangeName = eventTypeToPublisherName(eventType);
if (!(this.publisher[exchangeName] instanceof Function)) {
   .....
}
....
await this.publisher[exchangeName](msg);
```